Library/Homebrew/untap.rb

```diff
@@ -0,0 +1,56 @@
+# typed: true
+# frozen_string_literal: true
+
+module Homebrew
+  # Helpers for the `brew untap` command.
+  # @api private
+  module Untap
+    # All installed formulae currently available in a tap by formula full name.
+    sig { params(tap: Tap).returns(T::Array[Formula]) }
+    def self.installed_formulae_for(tap:)
+      tap.formula_names.filter_map do |formula_name|
+        next unless installed_formulae_names.include?(T.must(formula_name.split("/").last))
+
+        formula = begin
+          Formulary.factory(formula_name)
+        rescue
+          # Don't blow up because of a single unavailable formula.
+          next
+        end
+
+        # Can't use Formula#any_version_installed? because it doesn't consider
+        # taps correctly.
+        formula if formula.installed_kegs.any? { |keg| keg.tab.tap == tap }
+      end
+    end
+
+    sig { returns(T::Set[String]) }
+    def self.installed_formulae_names
+      @installed_formulae_names ||= Formula.installed_formula_names.to_set.freeze
```

Comment:
It may be used multiple times if there are multiple taps that get untapped at once. That probably doesn't happen that often but it is technically possible. Would using `cachable` here to make sure it gets cleared in tests be good enough? I'm not worried about it lasting the lifetime of the program because it's only used by one subcommand.

Reply:
> Would using `cachable` here to make sure it gets cleared in tests be good enough? 

Yes, either that, or pass it in from outside via a second parameter.

Alternatively, this could also be a class instead of a module, which you then instantiate once in `cmd/untap.rb` and multiple times in tests, making the cache more localised.

Reply:
Handled in https://github.com/Homebrew/brew/pull/16875/commits/eed7fd81a3fa3f4cbbdecaddbaaac03aa904d5f3